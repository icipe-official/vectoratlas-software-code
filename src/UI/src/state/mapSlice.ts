import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { fetchApiJson } from '../api/api';

export interface MapState {
  map_styles: {
    layers: {
      name: string;
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };

  map_overlays: {
    name: string;
    source: string;
    sourceType: string;
    layers: { name: string }[];
  }[];
}

export const initialState: MapState = {
  map_styles: { layers: [] },
  map_overlays: [],
};

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchApiJson('config/map-styles');
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchApiJson(
      'config/tile-server-overlays'
    );
    return tileServerOverlays;
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = action.payload;
      });
  },
});

export default mapSlice.reducer;
