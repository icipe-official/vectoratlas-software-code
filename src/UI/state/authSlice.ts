import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as njwt from 'njwt';
import { fetchAuth } from '../api/api';

export interface AuthState {
  roles: String[];
  token: String;
}

export const initialState: AuthState = {
  roles: [],
  token: '',
};

export const getUserInfo = createAsyncThunk('auth/getUserInfo', async () => {
  const token = await fetchAuth();
  const verifiedToken: any = njwt.verify(
    token,
    process.env.NEXT_PUBLIC_TOKEN_KEY
  );
  return {
    token: token,
    roles: verifiedToken?.body.scope.split(','),
  };
});

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getUserInfo.fulfilled, (state, action: any) => {
      state.roles = action.payload.roles;
      state.token = action.payload.token;
    });
  },
});

export default authSlice.reducer;
