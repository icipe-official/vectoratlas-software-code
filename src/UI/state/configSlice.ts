import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import {
  fetchLocalVersion,
  fetchApiVersion,
  fetchFeatureFlags,
} from '../api/api';

export interface ConfigState {
  version_ui: string;
  version_api: string;
  feature_flags_status: string;
  feature_flags: {
    flag: string;
    on: boolean;
  }[];
}

export const initialState: ConfigState = {
  version_ui: 'local_ui',
  version_api: 'local_api',
  feature_flags: [],
  feature_flags_status: '',
};

export const getUiVersion = createAsyncThunk(
  'config/getUiVersion',
  async () => {
    const version = await fetchLocalVersion();
    return version;
  }
);

export const getApiVersion = createAsyncThunk(
  'config/getApiVersion',
  async () => {
    const version = await fetchApiVersion();
    return version;
  }
);

export const getFeatureFlags = createAsyncThunk(
  'config/getFeatureFlags',
  async () => {
    const featureFlags = await fetchFeatureFlags();
    return featureFlags;
  }
);

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getUiVersion.pending, (state) => {
        state.version_ui = 'loading';
      })
      .addCase(getUiVersion.rejected, (state) => {
        state.version_ui = 'error';
      })
      .addCase(getUiVersion.fulfilled, (state, action) => {
        state.version_ui = action.payload;
      })
      .addCase(getApiVersion.pending, (state) => {
        state.version_api = 'loading';
      })
      .addCase(getApiVersion.rejected, (state) => {
        state.version_api = 'error';
      })
      .addCase(getApiVersion.fulfilled, (state, action) => {
        state.version_api = action.payload;
      })
      .addCase(getFeatureFlags.pending, (state) => {
        state.feature_flags_status = 'loading';
      })
      .addCase(getFeatureFlags.rejected, (state) => {
        state.feature_flags_status = 'error';
      })
      .addCase(getFeatureFlags.fulfilled, (state, action) => {
        state.feature_flags_status = 'success';
        state.feature_flags = action.payload;
      });
  },
});

export default configSlice.reducer;
