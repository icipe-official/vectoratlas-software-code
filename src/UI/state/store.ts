import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import rootReducer from './rootReducer';

export function makeStore() {
  return configureStore({
    reducer: rootReducer,
  });
}

const store = makeStore();

export type AppState = ReturnType<typeof store.getState>;
export const rootInitialState = store.getState() as AppState;

export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>;

export default store;
