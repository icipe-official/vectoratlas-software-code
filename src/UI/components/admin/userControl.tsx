import { IconButton } from '@mui/material';
import Checkbox from '@mui/material/Checkbox';
import Typography from '@mui/material/Typography';
import React, { useState } from 'react';
import {
  saveUserRoles,
} from '../../state/admin/actions/admin.actions';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { UsersWithRoles } from '../../state/state.types';
import SaveIcon from '@mui/icons-material/Save';

export const areRolesDifferent = (
  user: UsersWithRoles,
  workingCopy: UsersWithRoles
) => {
  return (
    user.is_admin !== workingCopy.is_admin ||
    user.is_uploader !== workingCopy.is_uploader ||
    user.is_reviewer !== workingCopy.is_reviewer ||
    user.is_editor !== workingCopy.is_editor
  );
};

export const UserControl = ({ user }: { user: UsersWithRoles }) => {
  const dispatch = useAppDispatch();
  const savingRoles = useAppSelector((s) => s.admin.savingUser);

  const [workingCopy, setWorkingCopy] = useState({ ...user });

  const onChange = (propName: string) => (e) => {
    setWorkingCopy({
      ...workingCopy,
      [propName]: e.target.checked,
    });
  };

  const onSave = () => {
    dispatch(saveUserRoles(workingCopy));
  };

  return (
    <div
      style={{
        display: 'flex',
        width: '100%',
        alignItems: 'center',
        marginBottom: '20px',
      }}
    >
      <div style={{ flexGrow: 1, alignItems: 'center' }}>
        <Typography style={{ fontWeight: 'bold', marginRight: '5px' }}>
          {user.email}
        </Typography>
        <Typography variant="caption">({user.auth0_id})</Typography>
      </div>
      <Typography style={{}}>Uploader:</Typography>
      <Checkbox
        checked={workingCopy.is_uploader}
        onChange={onChange('is_uploader')}
      />
      <Typography style={{}}>Reviewer:</Typography>
      <Checkbox
        checked={workingCopy.is_reviewer}
        onChange={onChange('is_reviewer')}
      />
      <Typography style={{}}>Editor:</Typography>
      <Checkbox
        checked={workingCopy.is_editor}
        onChange={onChange('is_editor')}
      />
      <Typography style={{}}>Admin:</Typography>
      <Checkbox
        checked={workingCopy.is_admin}
        onChange={onChange('is_admin')}
      />
      <IconButton
        disabled={savingRoles || !areRolesDifferent(user, workingCopy)}
        onClick={onSave}
      >
        <SaveIcon />
      </IconButton>
    </div>
  );
};