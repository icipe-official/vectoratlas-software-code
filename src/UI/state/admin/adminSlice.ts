import { createSlice } from '@reduxjs/toolkit';
import { updateUserRoles } from '../../api/queries';
import { UsersWithRoles } from '../state.types';

export interface AdminState {
  users: UsersWithRoles[];
  loading: boolean;
  savingUser: boolean;
}

export const initialState: () => AdminState = () => ({
  users: [],
  loading: false,
  savingUser: false
});

export const adminSlice = createSlice({
  name: 'news',
  initialState: initialState(),
  reducers: {
    usersWithRoles(state, action) {
      state.users = action.payload;
    },
    adminLoading(state, action) {
      state.loading = action.payload;
    },
    savingRoles(state, action) {
      state.savingUser = action.payload;
    },
    changeUserRoles(state, action) {
      const matchingUser = state.users.find(u => u.auth0_id === action.payload.auth0_id);
      if (matchingUser) {
        matchingUser.is_admin = action.payload.is_admin;
        matchingUser.is_uploader = action.payload.is_uploader;
        matchingUser.is_reviewer = action.payload.is_reviewer;
        matchingUser.is_editor = action.payload.is_editor;
      }
    }
  },
  extraReducers: () => {},
});

export const {
  usersWithRoles,
  adminLoading,
  savingRoles,
  changeUserRoles
} = adminSlice.actions;

export default adminSlice.reducer;
