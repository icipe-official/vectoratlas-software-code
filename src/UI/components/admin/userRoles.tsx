import CircularProgress from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import React, { useEffect } from 'react';
import { getUserRoles } from '../../state/admin/actions/admin.actions';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { UserControl } from './userControl';

export const UserRolePanel = () => {
  const dispatch = useAppDispatch();
  const loadingUserRoles = useAppSelector((s) => s.admin.loading);

  const userRoles = useAppSelector((s) => s.admin.users);

  useEffect(() => {
    dispatch(getUserRoles());
  }, [dispatch]);

  return (
    <div>
      {loadingUserRoles ? (
        <div
          style={{ width: '100%', display: 'flex', justifyContent: 'center' }}
        >
          <CircularProgress />
        </div>
      ) : null}
      <Typography
        color="primary"
        variant="h5"
        sx={{ mt: 2, mb: 1 }}
        style={{ flexGrow: 1 }}
      >
        User Roles
      </Typography>
      <div>
        {userRoles.map((u) => {
          return <UserControl key={u.auth0_id} user={u} />;
        })}
      </div>
    </div>
  );
};
