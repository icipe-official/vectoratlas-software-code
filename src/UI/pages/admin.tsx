import { Container, Typography } from '@mui/material';
import { UserRolePanel } from '../components/admin/userRoles';
import AuthWrapper from '../components/shared/AuthWrapper';

export default function SourcesPage(): JSX.Element {
  return (
    <div>
      <main>
        <Container>
          <AuthWrapper role="admin">
            <div>
              <Typography
                color="primary"
                variant="h4"
                sx={{ mt: 2, mb: 1 }}
                style={{ flexGrow: 1 }}
              >
                Administration
              </Typography>
              <UserRolePanel />
            </div>
          </AuthWrapper>
        </Container>
      </main>
    </div>
  );
}
