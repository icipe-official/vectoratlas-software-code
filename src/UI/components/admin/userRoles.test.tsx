import React from 'react';
import { render } from '../../test_config/render';
import { UserRolePanel } from './userRoles';
import { AppState } from '../../state/store';
import { initialState } from '../../state/admin/adminSlice';

jest.mock(
  '@mui/material/CircularProgress',
  () =>
    function CircularProgressMock() {
      return <div>Circular progress mock</div>;
    }
);

jest.mock('./userControl', () => ({
  UserControl: (props) => <div>User control mock {JSON.stringify(props)}</div>,
}));

jest.mock('../../state/admin/actions/admin.actions', () => ({
  getUserRoles: () => ({ type: 'getUserRoles-mock' }),
}));

describe('UserRolePanel', () => {
  let state: Partial<AppState>;

  beforeEach(() => {
    jest.clearAllMocks();

    state = {
      admin: initialState(),
    };
  });

  it('loads roles when intialised', () => {
    const { store } = render(<UserRolePanel />, state);

    const actions = store.getActions();
    expect(actions[0]).toEqual({ type: 'getUserRoles-mock' });
  });

  it('shows loading spinning when loading', () => {
    state.admin.loading = true;

    const { wrapper } = render(<UserRolePanel />, state);

    expect(wrapper.getByText('Circular progress mock')).toBeInTheDocument();
  });

  it('renders correctly with users', () => {
    state.admin.users = [
      {
        auth0_id: 'A',
        email: 'user A',
        is_admin: true,
        is_editor: true,
        is_uploader: false,
        is_reviewer: false,
      },
      {
        auth0_id: 'B',
        email: 'user B',
        is_admin: false,
        is_editor: false,
        is_uploader: true,
        is_reviewer: false,
      },
    ];

    const { wrapper } = render(<UserRolePanel />, state);
    expect(wrapper.container).toMatchSnapshot();
  });
});
