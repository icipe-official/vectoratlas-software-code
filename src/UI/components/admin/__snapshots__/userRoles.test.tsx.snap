// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRolePanel renders correctly with users 1`] = `
<div>
  <div>
    <h5
      class="MuiTypography-root MuiTypography-h5 css-1704fb8-MuiTypography-root"
      style="flex-grow: 1;"
    >
      User Roles
    </h5>
    <div>
      <div>
        User control mock 
        {"user":{"auth0_id":"A","email":"user A","is_admin":true,"is_editor":true,"is_uploader":false,"is_reviewer":false}}
      </div>
      <div>
        User control mock 
        {"user":{"auth0_id":"B","email":"user B","is_admin":false,"is_editor":false,"is_uploader":true,"is_reviewer":false}}
      </div>
    </div>
  </div>
</div>
`;
